### ISSUE TITLE
Rule proposal: Don't pass default values to functions

### ISSUE THREAD

**Issue by fregante** [typescript-eslint/typescript-eslint#2304]
# Fail

```js
function run(always = true) {}
run(true); // Error
```
```js
function run({always} = {always: true}) {}
run({always: true}); // Error
```
```js
function run({always = true} = {}) {}
run({always: true}); // Error
```
```js
function run(options) {
	options = {
		always: true,
		...options
	}
}
run({always: true}); // Error
```
```js
function run(options) {
	options = Object.assign({
		always: true
	}, options);
}
run({always: true}); // Error
```

And any other common defaults pattern (unless already blocked by other rules like [`prefer-object-spread`](https://eslint.org/docs/rules/prefer-object-spread))

# Pass

```js
run(); // No parameter set
run({always: false}); // Non-default parameter set
run({callback: fn}); // Any non-default parameter set
```

# Exclude

If `run` depends on arity/`arguments.length`

**Comment by bradzacher** [comment]
is there anything typescript specific here?
If not, then this would be better as a request to the core eslint module.

**Comment by fregante** [comment]
It's specific in that TypeScript can apply this rule across files. Eslint rejected it for this very reason

**Comment by bradzacher** [comment]
ahh I misunderstood, sorry. I misread and thought that this was saying "don't use default values", but it's "don't call a function with an argument value that's identical to the argument's default".

----

I think this is technically possible to do, but it will be really, really complicated to do and will likely be an inexact, inaccurate science.

TypeScript deals with and tracks types of things, not values.
To TypeScript, your function `run` has parameter `{always}`, which has type `{always: boolean}`.
When you call the function, you pass argument `{always: true}`, which has type `{always: boolean}`.

So you inspect the types and they'll look the same.
If you want to, you can instead inspecting TS symbol for the argument, and you can find the TS AST Node for the argument, and find it has an initialiser. From there you can manually traverse and compare a TS AST Node to TS AST Node.

However, if the argument or the default value have a variable assigned to them, then you'll have to manually attempt to find and figure out the value of the variable.
In the current file that's easy, because ESLint has the scope manager API - however in the other file where the function is likely defined, the rule will have to manually convert the TS AST to an ESTree AST, and then trigger scope analysis on the AST before you can use the tooling.

Even with the scope analyser though, you'll be unable to resolve things like variables declared across certain module boundaries (i.e. modules with an `@types` decl won't have an inspectable value), or function boundaries (i.e. in `function foo(arg: string) { return (arg2 = arg) => {} }` you can never know the true value of `arg`).

**Comment by JoshuaKGoldberg** [comment]
> but it will be really, really complicated to do and will likely be an inexact, inaccurate science.

Given that this issue has received almost no attention since being filed in 2020, closing as out of scope for us. I'd encourage anybody who really wants this to build it as a custom plugin. See https://typescript-eslint.io/developers/custom-rules.

Cheers, thanks thanks for the discussion all!
